File: tests/benchmarks/common/lua/benchmark_runner.lua
local M = {}

local MOD = 1000000007
local TIER_SCALES = {
    smoke = 1,
    core = 4,
    stress = 16,
    profile = 1,
}

local function parse_scale(args)
    local tier = "core"
    local explicit_scale = nil
    local index = 1

    while index <= #args do
        local arg = args[index]
        if arg == "--tier" then
            if index + 1 > #args then
                error("--tier requires smoke, core, stress, or profile")
            end
            tier = args[index + 1]
            index = index + 2
        elseif arg == "--scale" then
            local parsed_scale
            if index + 1 > #args then
                error("--scale requires a positive integer")
            end
            parsed_scale = tonumber(args[index + 1])
            if parsed_scale == nil or parsed_scale < 1 or math.floor(parsed_scale) ~= parsed_scale then
                error("--scale requires a positive integer")
            end
            explicit_scale = parsed_scale
            index = index + 2
        else
            error("unknown argument: " .. tostring(arg))
        end
    end

    if explicit_scale ~= nil then
        return explicit_scale
    end
    if TIER_SCALES[tier] == nil then
        error("unsupported tier: " .. tostring(tier))
    end
    return TIER_SCALES[tier]
end

local function numeric_loops(scale)
    local outer_limit = 24 * scale
    local inner_limit = 3000 * scale
    local value = 17
    local checksum = 0

    for outer = 0, outer_limit - 1 do
        for inner = 0, inner_limit - 1 do
            value = (value * 1103 + 97 + outer + inner) % 65521
            if value % 7 == 0 then
                checksum = checksum + math.floor(value / 7)
            elseif value % 5 == 0 then
                checksum = checksum + (value % 97)
            else
                checksum = checksum + (value % 31)
            end
            checksum = checksum % MOD
        end
        checksum = (checksum + outer * 17 + (value % 13)) % MOD
    end

    return checksum
end

local function int_xor(left, right)
    if type(bit32) == "table" and bit32.bxor ~= nil then
        return bit32.bxor(left, right)
    end

    local result = 0
    local bit = 1
    local a = left
    local b = right

    while a > 0 or b > 0 do
        local a_bit = a % 2
        local b_bit = b % 2
        if a_bit ~= b_bit then
            result = result + bit
        end
        a = math.floor(a / 2)
        b = math.floor(b / 2)
        bit = bit * 2
    end

    return result
end

local function worker_step(worker, delta)
    if worker.kind == 0 then
        worker.state = (worker.state * 13 + delta + 7) % 10007
    elseif worker.kind == 1 then
        worker.state = (worker.state * 17 + delta * 3 + 11) % 10009
    elseif worker.kind == 2 then
        worker.state = (int_xor(worker.state, delta + 31) + delta * 5 + 19) % 10037
    else
        worker.state = (worker.state + delta * delta + 23) % 10039
    end
    return worker.state
end

local function dispatch_loops(scale)
    local workers = {
        { state = 17, kind = 0 },
        { state = 29, kind = 1 },
        { state = 43, kind = 2 },
        { state = 61, kind = 3 },
    }
    local outer_limit = 120 * scale
    local inner_limit = 320 * scale
    local checksum = 0

    for outer = 0, outer_limit - 1 do
        for inner = 0, inner_limit - 1 do
            local index = ((outer + inner) % 4) + 1
            local delta = outer * 7 + inner * 11 + (index - 1)
            local value = worker_step(workers[index], delta)
            checksum = (checksum + value * index + (delta % 29)) % MOD
        end
        local worker = workers[(outer % 4) + 1]
        checksum = (checksum + worker.state * (outer + 1)) % MOD
    end

    return checksum
end

local function container_label(value)
    if value % 2 == 0 then
        return "even"
    end
    if value > 128 then
        return "odd_hi"
    end
    return "odd_lo"
end

local function container_pipeline(scale)
    local total = 1024 * scale
    local queue = {}
    local seen = {}
    local buckets = {
        even = {},
        odd_hi = {},
        odd_lo = {},
    }
    local seed = 41

    for index = 0, total - 1 do
        seed = (seed * 29 + 17 + index) % 257
        queue[#queue + 1] = { container_label(seed), seed * scale + (index % 13) }
    end

    for _, item in ipairs(queue) do
        local label = item[1]
        local value = item[2]
        local key = label .. ":" .. tostring(value)
        if seen[key] == nil then
            seen[key] = true
            buckets[label][#buckets[label] + 1] = value
        end
    end

    local odd_lo_sum = 0
    for _, value in ipairs(buckets.odd_lo) do
        odd_lo_sum = odd_lo_sum + value
    end
    local odd_hi_sum = 0
    for _, value in ipairs(buckets.odd_hi) do
        odd_hi_sum = odd_hi_sum + value
    end
    local even_sum = 0
    for _, value in ipairs(buckets.even) do
        even_sum = even_sum + value
    end

    local unique_count = 0
    for _ in pairs(seen) do
        unique_count = unique_count + 1
    end

    return (even_sum * 100000 + odd_hi_sum * 100 + odd_lo_sum + unique_count) % MOD
end

local function insertion_sort(values)
    for index = 2, #values do
        local key = values[index]
        local cursor = index - 1
        while cursor >= 1 and values[cursor] > key do
            values[cursor + 1] = values[cursor]
            cursor = cursor - 1
        end
        values[cursor + 1] = key
    end
end

local function mod_mul(left, right, modulus)
    local result = 0
    local a = left % modulus
    local b = right

    while b > 0 do
        if b % 2 == 1 then
            result = (result + a) % modulus
        end
        b = math.floor(b / 2)
        a = (a * 2) % modulus
    end

    return result
end

local function build_sort_pattern(pattern, length)
    local values = {}
    local seed = 97

    if pattern == 0 then
        for index = 0, length - 1 do
            seed = (mod_mul(seed, 1103515245, 2147483647) + 12345 + index) % 2147483647
            values[#values + 1] = seed % 100000
        end
        return values
    end

    if pattern == 1 then
        for index = 0, length - 1 do
            values[#values + 1] = length - index
        end
        return values
    end

    if pattern == 2 then
        for index = 0, length - 1 do
            values[#values + 1] = (index * 17 + 3) % (math.floor(length / 8) + 5)
        end
        return values
    end

    for index = 0, length - 1 do
        values[#values + 1] = index * 3 + (index % 7)
    end
    for index = 1, length, 7 do
        local swap_index = ((index - 1) * 13 + 5) % length + 1
        values[index], values[swap_index] = values[swap_index], values[index]
    end
    return values
end

local function sort_array(scale)
    local length = 16 * scale
    local step = math.floor(length / 7)
    local checksum = 0
    if step < 1 then
        step = 1
    end

    for pattern = 0, 3 do
        local values = build_sort_pattern(pattern, length)
        insertion_sort(values)
        local cursor = 1
        local subtotal = 0
        while cursor <= length do
            subtotal = (subtotal + values[cursor] * cursor) % MOD
            cursor = cursor + step
        end
        subtotal = (subtotal
            + values[1] * 3
            + values[math.floor(length / 2) + 1] * 5
            + values[length] * 7
            + pattern * 11) % MOD
        checksum = (checksum * 131 + subtotal) % MOD
    end

    return checksum
end

local function prime_trial_division(scale)
    local limit = 5000 * scale
    local checksum = 0
    local count = 0

    for candidate = 2, limit do
        local is_prime = true
        local divisor = 2
        while divisor * divisor <= candidate do
            if candidate % divisor == 0 then
                is_prime = false
                break
            end
            divisor = divisor + 1
        end

        if is_prime then
            count = count + 1
            checksum = (checksum + candidate * ((count % 97) + 1)) % MOD
        end
    end

    return checksum
end

local function matrix_add_2d(scale)
    local rows = 24 * scale
    local cols = 32 * scale
    local cells = rows * cols
    local lhs = {}
    local rhs = {}
    local dst = {}
    local scratch = {}
    local checksum = 0

    for index = 1, cells do
        lhs[index] = ((index - 1) * 13 + 7) % 997
        rhs[index] = ((index - 1) * 17 + 11) % 991
    end

    for row = 0, rows - 1 do
        local row_sum = 0
        for col = 0, cols - 1 do
            local index = row * cols + col + 1
            dst[index] = lhs[index] + rhs[index] + ((row + col) % 7)
            scratch[index] = dst[index] - math.floor(lhs[index] / 3) + (rhs[index] % 11)
            row_sum = row_sum + scratch[index] * (col + 1)
        end
        checksum = (checksum + row_sum * (row + 1)) % MOD
    end

    for index = 1, cells do
        checksum = (checksum + scratch[index] * (((index - 1) % 17) + 1)) % MOD
    end

    return checksum
end

local function string_build(scale)
    local fragments = { "al", "be", "cy", "do", "ex", "fu" }
    local counts = {}
    local keys = {}
    local assembled = ""
    local assembled_score = 0
    local checksum = 0
    local seed = 17
    local iterations = 180 * scale

    for index = 0, iterations - 1 do
        seed = (seed * 73 + 19 + index) % 997
        local token_id = (seed + index) % 23
        local token = fragments[(seed % #fragments) + 1]
            .. "-"
            .. fragments[(token_id % #fragments) + 1]
            .. fragments[((token_id + 2) % #fragments) + 1]
        local token_score = (seed % 211) + token_id * 17 + index
        assembled = assembled .. token
        assembled_score = (assembled_score * 41 + token_score) % MOD
        if index % 4 == 0 then
            assembled = assembled .. "|"
            assembled_score = (assembled_score + 3) % MOD
        else
            assembled = assembled .. ":"
            assembled_score = (assembled_score + 7) % MOD
        end

        if index % 9 == 8 then
            if counts[assembled] == nil then
                counts[assembled] = 0
                keys[#keys + 1] = assembled
            end
            counts[assembled] = (counts[assembled] + assembled_score + index + 1) % MOD
            checksum = (checksum + counts[assembled] + (seed % 97)) % MOD
            assembled = token
            assembled_score = token_score % MOD
        end
    end

    if #assembled > 0 then
        if counts[assembled] == nil then
            counts[assembled] = 0
            keys[#keys + 1] = assembled
        end
        counts[assembled] = (counts[assembled] + assembled_score + iterations) % MOD
    end

    for index = 1, #keys do
        checksum = (checksum + counts[keys[index]] * index) % MOD
    end

    return checksum
end

local function map_object_access(scale)
    local labels = { "aa", "bb", "cc", "dd" }
    local buckets = {}
    local checksum = 0
    local left = 3
    local right = 7
    local hits = 0
    local outer_limit = 64 * scale

    for outer = 0, outer_limit - 1 do
        for inner = 0, 31 do
            left = (left * 31 + outer + inner + hits) % 10007
            right = (right + left + inner * 3 + 5) % 10009
            hits = hits + 1
            local label = labels[((outer + inner) % #labels) + 1]
            local key = label .. "_slot"
            buckets[key] = ((buckets[key] or 0) + left + right + hits) % MOD
            checksum = (checksum + buckets[key] + left + hits) % MOD
        end
    end

    checksum = checksum
        + (buckets["aa_slot"] or 0)
        + (buckets["bb_slot"] or 0)
        + (buckets["cc_slot"] or 0)
        + (buckets["dd_slot"] or 0)
    return checksum % MOD
end

local CASE_HANDLERS = {
    numeric_loops = { "BENCH_NUMERIC_LOOPS_PASS", numeric_loops },
    dispatch_loops = { "BENCH_DISPATCH_LOOPS_PASS", dispatch_loops },
    container_pipeline = { "BENCH_CONTAINER_PIPELINE_PASS", container_pipeline },
    sort_array = { "BENCH_SORT_ARRAY_PASS", sort_array },
    prime_trial_division = { "BENCH_PRIME_TRIAL_DIVISION_PASS", prime_trial_division },
    matrix_add_2d = { "BENCH_MATRIX_ADD_2D_PASS", matrix_add_2d },
    string_build = { "BENCH_STRING_BUILD_PASS", string_build },
    map_object_access = { "BENCH_MAP_OBJECT_ACCESS_PASS", map_object_access },
}

function M.run_main(case_name, args)
    local handler = CASE_HANDLERS[case_name]
    if handler == nil then
        error("unknown benchmark case: " .. tostring(case_name))
    end

    local scale = parse_scale(args or {})
    print(handler[1])
    print(handler[2](scale))
end

return M

File: tests/benchmarks/common/qjs/benchmark_runner.js
const MOD = 1000000007;
const TIER_SCALES = {
    smoke: 1,
    core: 4,
    stress: 16,
    profile: 1,
};

function parseScale(args) {
    let tier = "core";
    let explicitScale = null;

    for (let index = 0; index < args.length; index += 1) {
        const arg = args[index];
        if (arg === "--tier") {
            if (index + 1 >= args.length) {
                throw new Error("--tier requires smoke, core, stress, or profile");
            }
            tier = args[index + 1];
            index += 1;
            continue;
        }
        if (arg === "--scale") {
            if (index + 1 >= args.length) {
                throw new Error("--scale requires a positive integer");
            }
            explicitScale = Number.parseInt(args[index + 1], 10);
            if (!Number.isFinite(explicitScale) || explicitScale < 1) {
                throw new Error("--scale requires a positive integer");
            }
            index += 1;
            continue;
        }
        throw new Error(`unknown argument: ${arg}`);
    }

    if (explicitScale !== null) {
        return explicitScale;
    }
    if (!Object.prototype.hasOwnProperty.call(TIER_SCALES, tier)) {
        throw new Error(`unsupported tier: ${tier}`);
    }
    return TIER_SCALES[tier];
}

function normalizeArgs(args) {
    if (!args || args.length === 0) {
        return [];
    }

    if (typeof args[0] === "string" && !args[0].startsWith("--") && args[0].endsWith(".js")) {
        return args.slice(1);
    }

    return args;
}

function numericLoops(scale) {
    const outerLimit = 24 * scale;
    const innerLimit = 3000 * scale;
    let value = 17;
    let checksum = 0;

    for (let outer = 0; outer < outerLimit; outer += 1) {
        for (let inner = 0; inner < innerLimit; inner += 1) {
            value = (value * 1103 + 97 + outer + inner) % 65521;
            if (value % 7 === 0) {
                checksum += Math.floor(value / 7);
            } else if (value % 5 === 0) {
                checksum += value % 97;
            } else {
                checksum += value % 31;
            }
            checksum %= MOD;
        }
        checksum = (checksum + outer * 17 + (value % 13)) % MOD;
    }

    return checksum;
}

class MultiplyWorker {
    constructor(seed) {
        this.state = seed;
    }

    step(delta) {
        this.state = (this.state * 13 + delta + 7) % 10007;
        return this.state;
    }

    read() {
        return this.state;
    }
}

class ScaleWorker {
    constructor(seed) {
        this.state = seed;
    }

    step(delta) {
        this.state = (this.state * 17 + delta * 3 + 11) % 10009;
        return this.state;
    }

    read() {
        return this.state;
    }
}

class XorWorker {
    constructor(seed) {
        this.state = seed;
    }

    step(delta) {
        this.state = ((this.state ^ (delta + 31)) + delta * 5 + 19) % 10037;
        return this.state;
    }

    read() {
        return this.state;
    }
}

class DriftWorker {
    constructor(seed) {
        this.state = seed;
    }

    step(delta) {
        this.state = (this.state + delta * delta + 23) % 10039;
        return this.state;
    }

    read() {
        return this.state;
    }
}

function dispatchLoops(scale) {
    const workers = [
        new MultiplyWorker(17),
        new ScaleWorker(29),
        new XorWorker(43),
        new DriftWorker(61),
    ];
    const outerLimit = 120 * scale;
    const innerLimit = 320 * scale;
    let checksum = 0;

    for (let outer = 0; outer < outerLimit; outer += 1) {
        for (let inner = 0; inner < innerLimit; inner += 1) {
            const index = (outer + inner) & 3;
            const delta = outer * 7 + inner * 11 + index;
            const value = workers[index].step(delta);
            checksum = (checksum + value * (index + 1) + (delta % 29)) % MOD;
        }

        checksum = (checksum + workers[outer & 3].read() * (outer + 1)) % MOD;
    }

    return checksum;
}

function containerLabel(value) {
    if (value % 2 === 0) {
        return "even";
    }
    if (value > 128) {
        return "odd_hi";
    }
    return "odd_lo";
}

function containerPipeline(scale) {
    const total = 1024 * scale;
    const queue = [];
    const seen = new Set();
    const buckets = new Map();
    let seed = 41;

    for (let index = 0; index < total; index += 1) {
        seed = (seed * 29 + 17 + index) % 257;
        queue.push([containerLabel(seed), seed * scale + (index % 13)]);
    }

    for (let index = 0; index < queue.length; index += 1) {
        const [label, value] = queue[index];
        seen.add(`${label}:${value}`);
    }

    for (const item of seen) {
        const separator = item.indexOf(":");
        const label = item.slice(0, separator);
        const value = Number(item.slice(separator + 1));
        if (!buckets.has(label)) {
            buckets.set(label, []);
        }
        buckets.get(label).push(value);
    }

    const oddLoSum = (buckets.get("odd_lo") || []).reduce((sum, value) => sum + value, 0);
    const oddHiSum = (buckets.get("odd_hi") || []).reduce((sum, value) => sum + value, 0);
    const evenSum = (buckets.get("even") || []).reduce((sum, value) => sum + value, 0);
    return (evenSum * 100000 + oddHiSum * 100 + oddLoSum + seen.size) % MOD;
}

function insertionSort(values) {
    for (let index = 1; index < values.length; index += 1) {
        const key = values[index];
        let cursor = index - 1;
        while (cursor >= 0 && values[cursor] > key) {
            values[cursor + 1] = values[cursor];
            cursor -= 1;
        }
        values[cursor + 1] = key;
    }
}

function buildSortPattern(pattern, length) {
    const values = [];
    let seed = 97n;

    if (pattern === 0) {
        for (let index = 0; index < length; index += 1) {
            seed = (seed * 1103515245n + 12345n + BigInt(index)) % 2147483647n;
            values.push(Number(seed % 100000n));
        }
        return values;
    }

    if (pattern === 1) {
        for (let index = 0; index < length; index += 1) {
            values.push(length - index);
        }
        return values;
    }

    if (pattern === 2) {
        for (let index = 0; index < length; index += 1) {
            values.push((index * 17 + 3) % (Math.floor(length / 8) + 5));
        }
        return values;
    }

    for (let index = 0; index < length; index += 1) {
        values.push(index * 3 + (index % 7));
    }
    for (let index = 0; index < length; index += 7) {
        const swapIndex = (index * 13 + 5) % length;
        const temporary = values[index];
        values[index] = values[swapIndex];
        values[swapIndex] = temporary;
    }
    return values;
}

function sortArray(scale) {
    const length = 16 * scale;
    let step = Math.floor(length / 7);
    let checksum = 0;
    if (step < 1) {
        step = 1;
    }

    for (let pattern = 0; pattern < 4; pattern += 1) {
        const values = buildSortPattern(pattern, length);
        insertionSort(values);
        let cursor = 0;
        let subtotal = 0;
        while (cursor < length) {
            subtotal = (subtotal + values[cursor] * (cursor + 1)) % MOD;
            cursor += step;
        }
        subtotal = (subtotal
            + values[0] * 3
            + values[Math.floor(length / 2)] * 5
            + values[length - 1] * 7
            + pattern * 11) % MOD;
        checksum = (checksum * 131 + subtotal) % MOD;
    }

    return checksum;
}

function primeTrialDivision(scale) {
    const limit = 5000 * scale;
    let checksum = 0;
    let count = 0;

    for (let candidate = 2; candidate <= limit; candidate += 1) {
        let isPrime = true;
        for (let divisor = 2; divisor * divisor <= candidate; divisor += 1) {
            if (candidate % divisor === 0) {
                isPrime = false;
                break;
            }
        }
        if (isPrime) {
            count += 1;
            checksum = (checksum + candidate * ((count % 97) + 1)) % MOD;
        }
    }

    return checksum;
}

function matrixAdd2d(scale) {
    const rows = 24 * scale;
    const cols = 32 * scale;
    const cells = rows * cols;
    const lhs = new Array(cells).fill(0);
    const rhs = new Array(cells).fill(0);
    const dst = new Array(cells).fill(0);
    const scratch = new Array(cells).fill(0);
    let checksum = 0;

    for (let index = 0; index < cells; index += 1) {
        lhs[index] = (index * 13 + 7) % 997;
        rhs[index] = (index * 17 + 11) % 991;
    }

    for (let row = 0; row < rows; row += 1) {
        let rowSum = 0;
        for (let col = 0; col < cols; col += 1) {
            const index = row * cols + col;
            dst[index] = lhs[index] + rhs[index] + ((row + col) % 7);
            scratch[index] = dst[index] - Math.floor(lhs[index] / 3) + (rhs[index] % 11);
            rowSum += scratch[index] * (col + 1);
        }
        checksum = (checksum + rowSum * (row + 1)) % MOD;
    }

    for (let index = 0; index < cells; index += 1) {
        checksum = (checksum + scratch[index] * ((index % 17) + 1)) % MOD;
    }

    return checksum;
}

function stringBuild(scale) {
    const fragments = ["al", "be", "cy", "do", "ex", "fu"];
    const counts = new Map();
    const keys = [];
    let assembled = "";
    let assembledScore = 0;
    let checksum = 0;
    let seed = 17;
    const iterations = 180 * scale;

    for (let index = 0; index < iterations; index += 1) {
        seed = (seed * 73 + 19 + index) % 997;
        const tokenId = (seed + index) % 23;
        const token = fragments[seed % fragments.length]
            + "-"
            + fragments[tokenId % fragments.length]
            + fragments[(tokenId + 2) % fragments.length];
        const tokenScore = (seed % 211) + tokenId * 17 + index;
        assembled += token;
        assembledScore = (assembledScore * 41 + tokenScore) % MOD;
        if (index % 4 === 0) {
            assembled += "|";
            assembledScore = (assembledScore + 3) % MOD;
        } else {
            assembled += ":";
            assembledScore = (assembledScore + 7) % MOD;
        }

        if (index % 9 === 8) {
            if (!counts.has(assembled)) {
                counts.set(assembled, 0);
                keys.push(assembled);
            }
            counts.set(assembled, (counts.get(assembled) + assembledScore + index + 1) % MOD);
            checksum = (checksum + counts.get(assembled) + (seed % 97)) % MOD;
            assembled = token;
            assembledScore = tokenScore % MOD;
        }
    }

    if (assembled.length > 0) {
        if (!counts.has(assembled)) {
            counts.set(assembled, 0);
            keys.push(assembled);
        }
        counts.set(assembled, (counts.get(assembled) + assembledScore + iterations) % MOD);
    }

    for (let index = 0; index < keys.length; index += 1) {
        checksum = (checksum + counts.get(keys[index]) * (index + 1)) % MOD;
    }

    return checksum;
}

function mapObjectAccess(scale) {
    const labels = ["aa", "bb", "cc", "dd"];
    const buckets = new Map();
    let checksum = 0;
    let left = 3;
    let right = 7;
    let hits = 0;
    const outerLimit = 64 * scale;

    for (let outer = 0; outer < outerLimit; outer += 1) {
        for (let inner = 0; inner < 32; inner += 1) {
            left = (left * 31 + outer + inner + hits) % 10007;
            right = (right + left + inner * 3 + 5) % 10009;
            hits += 1;
            const label = labels[(outer + inner) % labels.length];
            const key = `${label}_slot`;
            buckets.set(key, ((buckets.get(key) || 0) + left + right + hits) % MOD);
            checksum = (checksum + buckets.get(key) + left + hits) % MOD;
        }
    }

    return (checksum
        + buckets.get("aa_slot")
        + buckets.get("bb_slot")
        + buckets.get("cc_slot")
        + buckets.get("dd_slot")) % MOD;
}

const CASE_HANDLERS = {
    numeric_loops: ["BENCH_NUMERIC_LOOPS_PASS", numericLoops],
    dispatch_loops: ["BENCH_DISPATCH_LOOPS_PASS", dispatchLoops],
    container_pipeline: ["BENCH_CONTAINER_PIPELINE_PASS", containerPipeline],
    sort_array: ["BENCH_SORT_ARRAY_PASS", sortArray],
    prime_trial_division: ["BENCH_PRIME_TRIAL_DIVISION_PASS", primeTrialDivision],
    matrix_add_2d: ["BENCH_MATRIX_ADD_2D_PASS", matrixAdd2d],
    string_build: ["BENCH_STRING_BUILD_PASS", stringBuild],
    map_object_access: ["BENCH_MAP_OBJECT_ACCESS_PASS", mapObjectAccess],
};

export function runMain(caseName, args) {
    if (!Object.prototype.hasOwnProperty.call(CASE_HANDLERS, caseName)) {
        throw new Error(`unknown benchmark case: ${caseName}`);
    }

    const [banner, handler] = CASE_HANDLERS[caseName];
    const scale = parseScale(normalizeArgs(args || []));
    const checksum = handler(scale);
    print(banner);
    print(String(checksum));
}
